File: frontend/src/api/quotes.ts
import { API_BASE_URL, apiFetch, getAccessToken } from "./client";

async function apiUpload<T>(path: string, formData: FormData): Promise<T> {
  const headers = new Headers();
  const token = getAccessToken();
  if (token) headers.set("Authorization", `Bearer ${token}`);
  const stationId = localStorage.getItem("active_station_id");
  if (stationId) headers.set("X-Station-Id", stationId);

  const response = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
    headers,
    body: formData,
    credentials: "include",
  });

  if (!response.ok) {
    let message = "Falha na requisição.";
    try {
      const payload = await response.json();
      message = payload?.error?.message ?? message;
    } catch {
      // ignore
    }
    throw new Error(message);
  }

  return response.json() as Promise<T>;
}

export type QuoteListItem = {
  id: string;
  quote_number: number;
  station_id: string;
  distributor_id: string;
  quoted_at: string;
  valid_until: string;
  source_channel: string;
  status: string;
  effective_status: string;
  version: number;
  item_count: number;
};

export type QuoteList = {
  items: QuoteListItem[];
  total: number;
  page: number;
  page_size: number;
  summary: Record<string, number>;
};

export type QuoteItem = {
  id: string;
  product_id: string;
  distribution_base_id: string | null;
  sequence: number;
  quoted_price_per_liter: string;
  payment_term_id: string;
  payment_type_snapshot: string;
  payment_term_days_snapshot: number;
  payment_term_name_snapshot: string;
  freight_type: string;
  freight_calculation_type: string;
  freight_value_total: string | null;
  freight_value_per_liter: string | null;
  discount_per_liter: string;
  rebate_per_liter: string;
  other_cost_per_liter: string;
  other_cost_description: string | null;
  minimum_volume_liters: string;
  available_volume_liters: string | null;
  delivery_expected_at: string | null;
  valid_until: string | null;
  notes: string | null;
  item_effective_status?: string;
  effective_valid_until?: string;
};

export type QuoteEvidence = {
  id: string;
  category: string;
  original_file_name: string;
  content_type: string;
  file_extension: string;
  size_bytes: number;
  sha256: string;
  is_supplemental: boolean;
  active: boolean;
  uploaded_by: string;
  uploaded_at: string;
};

export type Quote = {
  id: string;
  organization_id: string;
  station_id: string;
  distributor_id: string;
  distribution_base_id: string | null;
  quote_number: number;
  quoted_at: string;
  valid_until: string;
  source_channel: string;
  entry_method: string;
  seller_name: string | null;
  seller_contact: string | null;
  external_reference: string | null;
  source_description: string | null;
  notes: string | null;
  status: string;
  effective_status: string;
  version: number;
  replaces_quote_id: string | null;
  duplicated_from_quote_id: string | null;
  activated_at: string | null;
  items: QuoteItem[];
  evidences: QuoteEvidence[];
  warnings: string[];
};

export type QuoteHistoryEntry = {
  id: string;
  action: string;
  version: number;
  reason: string | null;
  changed_fields: Record<string, unknown> | null;
  metadata: Record<string, unknown> | null;
  user_id: string | null;
  created_at: string;
};

type QueryParams = Record<string, string | number | boolean | undefined>;

function buildQuery(params: QueryParams = {}) {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== "") search.set(key, String(value));
  });
  const query = search.toString();
  return query ? `?${query}` : "";
}

export async function fetchQuotes(params: QueryParams = {}) {
  return apiFetch<QuoteList>(`/api/v1/quotes${buildQuery(params)}`);
}

export async function fetchQuote(quoteId: string) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}`);
}

export async function createQuote(payload: Record<string, unknown>) {
  return apiFetch<Quote>("/api/v1/quotes", { method: "POST", body: JSON.stringify(payload) });
}

export async function updateQuote(quoteId: string, payload: Record<string, unknown>) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function addQuoteItem(quoteId: string, payload: Record<string, unknown>) {
  return apiFetch<QuoteItem>(`/api/v1/quotes/${quoteId}/items`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function activateQuote(quoteId: string, expectedVersion: number) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}/activate`, {
    method: "POST",
    body: JSON.stringify({ expected_version: expectedVersion }),
  });
}

export async function cancelQuote(quoteId: string, expectedVersion: number, reason: string) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}/cancel`, {
    method: "POST",
    body: JSON.stringify({ expected_version: expectedVersion, reason }),
  });
}

export async function reviseQuote(quoteId: string, reason: string) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}/revise`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  });
}

export async function duplicateQuote(quoteId: string, payload: Record<string, unknown>) {
  return apiFetch<Quote>(`/api/v1/quotes/${quoteId}/duplicate`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function fetchQuoteHistory(quoteId: string, page = 1) {
  return apiFetch<{ items: QuoteHistoryEntry[]; total: number }>(
    `/api/v1/quotes/${quoteId}/history${buildQuery({ page })}`,
  );
}

export async function uploadQuoteEvidence(
  quoteId: string,
  file: File,
  category: string,
  expectedVersion: number,
  isSupplemental = false,
) {
  const form = new FormData();
  form.append("file", file);
  form.append("category", category);
  form.append("expected_version", String(expectedVersion));
  form.append("is_supplemental", String(isSupplemental));
  return apiUpload<Quote>(`/api/v1/quotes/${quoteId}/evidences`, form);
}

export async function fetchItemPrefill(params: {
  station_id: string;
  distributor_id: string;
  product_id: string;
  distribution_base_id?: string;
}) {
  return apiFetch<{
    minimum_volume_liters: string;
    distribution_base_id: string | null;
    supplier_allowed: boolean;
    rule_source: string;
    alert_supplier_not_allowed: boolean;
  }>(`/api/v1/quotes/item-prefill${buildQuery(params)}`);
}

File: backend/app/schemas/quotes.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class QuoteItemBase(BaseModel):
    product_id: uuid.UUID
    distribution_base_id: uuid.UUID | None = None
    sequence: int | None = None
    quoted_price_per_liter: str
    payment_term_id: uuid.UUID
    freight_type: str = "CIF"
    freight_calculation_type: str = "NONE"
    freight_value_total: str | None = None
    freight_value_per_liter: str | None = None
    discount_per_liter: str = "0.0000"
    rebate_per_liter: str = "0.0000"
    other_cost_per_liter: str = "0.0000"
    other_cost_description: str | None = None
    minimum_volume_liters: str
    available_volume_liters: str | None = None
    delivery_expected_at: datetime | None = None
    valid_until: datetime | None = None
    notes: str | None = None


class QuoteItemCreate(QuoteItemBase):
    expected_version: int


class QuoteItemUpdate(BaseModel):
    expected_version: int
    product_id: uuid.UUID | None = None
    distribution_base_id: uuid.UUID | None = None
    sequence: int | None = None
    quoted_price_per_liter: str | None = None
    payment_term_id: uuid.UUID | None = None
    freight_type: str | None = None
    freight_calculation_type: str | None = None
    freight_value_total: str | None = None
    freight_value_per_liter: str | None = None
    discount_per_liter: str | None = None
    rebate_per_liter: str | None = None
    other_cost_per_liter: str | None = None
    other_cost_description: str | None = None
    minimum_volume_liters: str | None = None
    available_volume_liters: str | None = None
    delivery_expected_at: datetime | None = None
    valid_until: datetime | None = None
    notes: str | None = None


class QuoteItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    distribution_base_id: uuid.UUID | None
    sequence: int
    quoted_price_per_liter: str
    payment_term_id: uuid.UUID
    payment_type_snapshot: str
    payment_term_days_snapshot: int
    payment_term_name_snapshot: str
    freight_type: str
    freight_calculation_type: str
    freight_value_total: str | None
    freight_value_per_liter: str | None
    discount_per_liter: str
    rebate_per_liter: str
    other_cost_per_liter: str
    other_cost_description: str | None
    minimum_volume_liters: str
    available_volume_liters: str | None
    delivery_expected_at: datetime | None
    valid_until: datetime | None
    notes: str | None
    item_effective_status: str | None = None
    effective_valid_until: datetime | None = None

    @field_validator(
        "quoted_price_per_liter",
        "freight_value_total",
        "freight_value_per_liter",
        "discount_per_liter",
        "rebate_per_liter",
        "other_cost_per_liter",
        "minimum_volume_liters",
        "available_volume_liters",
        mode="before",
    )
    @classmethod
    def _decimal_to_str(cls, value: object) -> str | None:
        if value is None:
            return None
        if isinstance(value, Decimal):
            return str(value)
        return str(value)


class QuoteEvidenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    category: str
    original_file_name: str
    content_type: str
    file_extension: str
    size_bytes: int
    sha256: str
    is_supplemental: bool
    active: bool
    uploaded_by: uuid.UUID
    uploaded_at: datetime


class QuoteCreate(BaseModel):
    station_id: uuid.UUID
    distributor_id: uuid.UUID
    distribution_base_id: uuid.UUID | None = None
    quoted_at: datetime
    valid_until: datetime
    source_channel: str
    entry_method: str = "MANUAL"
    origin: str = "MANUAL_OPERATIONAL"
    analytics_eligible: bool = True
    seller_name: str | None = None
    seller_contact: str | None = None
    external_reference: str | None = None
    source_description: str | None = None
    notes: str | None = None


class QuoteUpdate(BaseModel):
    expected_version: int
    station_id: uuid.UUID | None = None
    distributor_id: uuid.UUID | None = None
    distribution_base_id: uuid.UUID | None = None
    quoted_at: datetime | None = None
    valid_until: datetime | None = None
    source_channel: str | None = None
    entry_method: str | None = None
    origin: str | None = None
    analytics_eligible: bool | None = None
    seller_name: str | None = None
    seller_contact: str | None = None
    external_reference: str | None = None
    source_description: str | None = None
    notes: str | None = None


class QuoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    station_id: uuid.UUID
    distributor_id: uuid.UUID
    distribution_base_id: uuid.UUID | None
    quote_number: int
    quoted_at: datetime
    valid_until: datetime
    source_channel: str
    entry_method: str
    origin: str
    analytics_eligible: bool
    seller_name: str | None
    seller_contact: str | None
    external_reference: str | None
    source_description: str | None
    notes: str | None
    status: str
    effective_status: str
    version: int
    replaces_quote_id: uuid.UUID | None
    duplicated_from_quote_id: uuid.UUID | None
    activated_at: datetime | None
    activated_by: uuid.UUID | None
    cancelled_at: datetime | None
    cancelled_by: uuid.UUID | None
    cancellation_reason: str | None
    superseded_at: datetime | None
    superseded_by_quote_id: uuid.UUID | None
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    items: list[QuoteItemResponse] = Field(default_factory=list)
    evidences: list[QuoteEvidenceResponse] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class QuoteListItem(BaseModel):
    id: uuid.UUID
    quote_number: int
    station_id: uuid.UUID
    distributor_id: uuid.UUID
    quoted_at: datetime
    valid_until: datetime
    source_channel: str
    status: str
    effective_status: str
    version: int
    item_count: int


class QuoteListResponse(BaseModel):
    items: list[QuoteListItem]
    total: int
    page: int
    page_size: int
    summary: dict[str, int]


class VersionedAction(BaseModel):
    expected_version: int


class CancelQuoteRequest(VersionedAction):
    reason: str


class ReviseQuoteRequest(BaseModel):
    reason: str


class DuplicateQuoteRequest(BaseModel):
    target_station_id: uuid.UUID
    quoted_at: datetime
    valid_until: datetime
    copy_evidences: bool = False
    notes: str | None = None


class DeactivateEvidenceRequest(BaseModel):
    reason: str


class QuoteHistoryEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    action: str
    version: int
    reason: str | None
    changed_fields: dict | None
    metadata: dict | None = Field(default=None, validation_alias="metadata_")
    user_id: uuid.UUID | None
    request_id: uuid.UUID | None
    created_at: datetime
    quote_item_id: uuid.UUID | None
    quote_evidence_id: uuid.UUID | None


class QuoteHistoryListResponse(BaseModel):
    items: list[QuoteHistoryEntry]
    total: int
    page: int
    page_size: int


class ItemPrefillResponse(BaseModel):
    minimum_volume_liters: str
    distribution_base_id: str | None
    supplier_allowed: bool
    rule_source: str
    alert_supplier_not_allowed: bool


class ExpirationRunResponse(BaseModel):
    expired_count: int
    skipped: bool = False
    reason: str | None = None
